src/coreclr/src/jit/gentree.h

```diff
@@ -5991,23 +5991,35 @@ struct GenTreePhiArg : public GenTreeLclVarCommon
 
 struct GenTreePutArgStk : public GenTreeUnOp
 {
+private:
+    unsigned m_byteOffset;
+#ifdef FEATURE_PUT_STRUCT_ARG_STK
+    unsigned m_byteSize; // The number of bytes that this argument is occupying on the stack with padding.
+#endif
+
+public:
+#if defined(DEBUG_ARG_SLOTS)
     unsigned gtSlotNum; // Slot number of the argument to be passed on stack
+#endif
 #if defined(UNIX_X86_ABI)
     unsigned gtPadAlign; // Number of padding slots for stack alignment
 #endif
 
     // Don't let clang-format mess with the GenTreePutArgStk constructor.
     // clang-format off
 
-    GenTreePutArgStk(genTreeOps   oper,
-                     var_types    type,
-                     GenTree*   op1,
-                     unsigned     slotNum
-                     PUT_STRUCT_ARG_STK_ONLY_ARG(unsigned numSlots),
-                     bool         putInIncomingArgArea = false,
-                     GenTreeCall* callNode = nullptr)
+    GenTreePutArgStk(genTreeOps   oper, var_types type, GenTree* op1, unsigned stackByteOffset
+                     PUT_STRUCT_ARG_STK_ONLY_ARG(unsigned stackByteSize)
+#if defined(DEBUG_ARG_SLOTS)
+        , unsigned slotNum
+        PUT_STRUCT_ARG_STK_ONLY_ARG(unsigned numSlots)
+#endif
+                     ,bool         putInIncomingArgArea = false
+                     ,GenTreeCall* callNode = nullptr)
         : GenTreeUnOp(oper, type, op1 DEBUGARG(/*largeNode*/ false))
-        , gtSlotNum(slotNum)
+        , m_byteOffset(stackByteOffset)
+        PUT_STRUCT_ARG_STK_ONLY_ARG(m_byteSize(stackByteSize))
+        DEBUG_ARG_SLOTS_ARG(gtSlotNum(slotNum))
```

Comment:
IMO this is now practically unreadable. I think it would be better to simply have different declarations for the different cases.

Reply:
yeah, I was moving it back and forth for a  while but was not able to make it "clear". I have changed it to different declarations, it looks worse as a diff, but probably more readable as a code block.

It was easy to lose an argument/field init, so I moved all fields to the beginning of the struct.

Reply:
Right - the duplication is unfortunate, but better, I think, than all the ifdefs and conditionally defined macros.